web3/_utils/normalizers.py

```diff
@@ -61,79 +65,103 @@ def wrapper(abi_type, data):
 
 
 @implicitly_identity
-def addresses_checksummed(abi_type, data):
-    if abi_type == 'address':
-        return abi_type, to_checksum_address(data)
+def addresses_checksummed(type_str, data):
+    if type_str == 'address':
+        return type_str, to_checksum_address(data)
 
 
 @implicitly_identity
-def decode_abi_strings(abi_type, data):
-    if abi_type == 'string':
-        return abi_type, codecs.decode(data, 'utf8', 'backslashreplace')
+def decode_abi_strings(type_str, data):
+    if type_str == 'string':
+        return type_str, codecs.decode(data, 'utf8', 'backslashreplace')
 
 
 #
 # Argument Normalizers
 #
 
 
+def parse_basic_type_str(old_normalizer):
+    """
+    Modifies a normalizer to automatically parse the incoming type string.  If
+    that type string does not represent a basic type (i.e. non-tuple type) or is
+    not parsable, the normalizer does nothing.
+    """
+    @functools.wraps(old_normalizer)
+    def new_normalizer(type_str, data):
+        try:
+            abi_type = parse(type_str)
+        except ParseError:
+            # If type string is not parsable, do nothing
+            return type_str, data
+
+        if not isinstance(abi_type, BasicType):
+            return type_str, data
+
+        return old_normalizer(abi_type, type_str, data)
+
+    return new_normalizer
+
+
 @implicitly_identity
-def abi_bytes_to_hex(abi_type, data):
-    base, sub, arrlist = process_type(abi_type)
-    if base == 'bytes' and not arrlist:
-        bytes_data = hexstr_if_str(to_bytes, data)
-        if not sub:
-            return abi_type, to_hex(bytes_data)
-        else:
-            num_bytes = int(sub)
-            if len(bytes_data) <= num_bytes:
-                padded = bytes_data.ljust(num_bytes, b'\0')
-                return abi_type, to_hex(padded)
-            else:
-                raise ValueError(
-                    "This value was expected to be at most %d bytes, but instead was %d: %r" % (
-                        (num_bytes, len(bytes_data), data)
-                    )
-                )
+@parse_basic_type_str
+def abi_bytes_to_hex(abi_type, type_str, data):
+    if abi_type.base != 'bytes' or abi_type.is_array:
+        return
+
+    bytes_data = hexstr_if_str(to_bytes, data)
+    if abi_type.sub is None:
+        return type_str, to_hex(bytes_data)
+
+    num_bytes = abi_type.sub
```

Comment:
I've preferred the style that is to keep the else block rather than have the `return` or `raise` imply the else block, and I think it's somewhat consistent elsewhere. I think it's nice to see the "shape" of the method, quickly. Instead of having to read it sequentially to build the possible code paths, they are laid out as indented.

It's less significant to me at the very beginning of a method, since it's so common to have a method short-circuit right up front.

I see now that it's not in the style guide, maybe I'll take a pass at adding it.